scheduling/allocator.rb

```diff
@@ -0,0 +1,239 @@
+      storage_volumes.size.times.zip(storage_volumes).to_h.sort_by { |k, v| v["size_gib"] * -1 },
+      distinct_storage_devices,
+      target_utilization,
+      vm.arch,
+      allocation_state_filter,
+      host_filter,
+      location_filter,
+      location_preference
+    )
+    allocation = Allocation.best_allocation(request)
+    fail "#{vm} no space left on any eligible host" unless allocation
+
+    allocation.update(vm)
+    Clog.emit("vm allocated") { {allocation: allocation.to_s, duration: Time.now - vm.created_at} }
+  end
+
+  Request = Struct.new(:vm_id, :cores, :mem_gib, :storage_gib, :storage_volumes, :distinct_storage_devices, :target_utilization,
+    :arch_filter, :allocation_state_filter, :host_filter, :location_filter, :location_preference)
+
+  class Allocation
+    attr_reader :score
+
+    def self.best_allocation(request)
+      candidate_hosts(request).map { Allocation.new(_1, request) }
+        .select { _1.is_valid }
+        .min_by { _1.score }
+    end
+
+    def self.candidate_hosts(request)
+      ds = DB[:vm_host].join(:storage_device, vm_host_id: :id)
+        .select_group(:vm_host_id, :total_cores, :used_cores, :total_hugepages_1g, :used_hugepages_1g, :location)
+        .select_append { count.function.*.as(num_storage_devices) }
+        .select_append { sum(available_storage_gib).as(available_storage_gib) }
+        .select_append { sum(total_storage_gib).as(total_storage_gib) }
+        .select_append { json_agg(json_build_object(Sequel.lit("'id'"), Sequel[:storage_device][:id], Sequel.lit("'total_storage_gib'"), total_storage_gib, Sequel.lit("'available_storage_gib'"), available_storage_gib)).order(available_storage_gib).as(storage_devices) }
+        .where(enabled: true)
+        .where(allocation_state: request.allocation_state_filter)
+        .where(arch: request.arch_filter)
+        .where { (total_hugepages_1g - used_hugepages_1g >= request.mem_gib) }
+        .where { (total_cores - used_cores >= request.cores) }
+        .having { sum(available_storage_gib) >= request.storage_gib }
+        .having { count.function.* >= (request.distinct_storage_devices ? request.storage_volumes.count : 1) }
+
+      ds = ds.where(vm_host_id: request.host_filter) unless request.host_filter.empty?
+      ds = ds.where(location: request.location_filter) unless request.location_filter.empty?
+      ds.all
+    end
+
+    def self.update_vm(vm_host, vm)
+      ip4, address = vm_host.ip4_random_vm_network if vm.ip4_enabled
+      fail "no ip4 addresses left" if vm.ip4_enabled && !ip4
+      vm.update(
+        vm_host_id: vm_host.id,
+        ephemeral_net6: vm_host.ip6_random_vm_network.to_s,
+        local_vetho_ip: vm_host.veth_pair_random_ip4_addr.to_s,
+        allocated_at: Time.now
+      )
+      AssignedVmAddress.create_with_id(dst_vm_id: vm.id, ip: ip4.to_s, address_id: address.id) if ip4
+      vm.sshable&.update(host: vm.ephemeral_net4 || vm.ephemeral_net6.nth(2))
```

Comment:
For now, updating the sshable with `ephemeral_net6` is useless since the production environment does not support `IPv6`. Therefore, any service that fully depends on IPv6 will not be able to make use of `sshable`. This is not a problem of this PR but writing for information sharing purposes. You may also consider removing it, it's upto you.

Reply:
let's leave it for now then